pyro/distributions/torch_patch.py

```diff
@@ -33,6 +33,57 @@ def decorator(new_fn):
     return decorator
 
 
+# See: tests.distributions.transforms.test_transforms.py::test_transform_backward
+# TODO: should be moved to PyTorch if this is a reasonable pattern
+@patch_dependency('torch.distributions.transforms.Transform.__call__')
+def _Transform__call__(self, x):
+    if self._cache_size == 0:
+        return self._call(x)
+    x_old, y_old = self._cached_x_y
+    if x is x_old:
+        return y_old
+    y = self._call(x)
+
+    if self._cache_size > 0 and x.requires_grad:
```

Comment:
How about restricting to leaf variables, since only they would be updated in-place?
```py
if self._cache_size > 0 and x.requires_grad and x.is_leaf:
```